[CIR] Generate `cir.dyn_cast` for dynamic casts to void ptr

Lancern opened the pull request:
This patch update the CIRGen of `dynamic_cast` expressions and make it start to generate `cir.dyn_cast` operations for `dynamic_cast` expressions that cast to a void pointer. This patch also updates the lowering prepare pass so that it lowers such `cir.dyn_cast` operations to the code emitted before this patch.

This patch adds a new attribute `#cir.dyn_cast_to_void` that carries ABI- specific information about a dynamic-cast-to-void-pointer operation. For downcasts and sidecasts, the `cir.dyn_cast` operation continues to hold a `#cir.dyn_cast_info` attribute. For dynamic-cast-to-void-pointer, the `cir.dyn_cast` operation holds a `#cir.dyn_cast_to_void` attribute instead.

bcardosolopes commented on clang/include/clang/CIR/Dialect/IR/CIRAttrs.td:
```diff
@@ -554,6 +554,36 @@ def StructLayoutAttr : CIR_Attr<"StructLayout", "struct_layout"> {
+
+  let assemblyFormat = [{
+    `<`
+      `vtableElemTy` `=` qualified($vtableElemTy) `,`
```
As a general approach, names printed in CIR asm should never have this type of case, I found one the other day that I missed during review and remove it, please let's not introduce additional ones. So something like this should be instead `vtable_elt_type`.

Lancern replied:
Updated.

bcardosolopes commented on clang/lib/CIR/CodeGen/CIRGenItaniumCXXABI.cpp:
```diff
@@ -2288,6 +2288,23 @@ static mlir::cir::FuncOp getItaniumDynamicCastFn(CIRGenFunction &CGF) {
+mlir::cir::DynamicCastToVoidInfoAttr
+CIRGenItaniumCXXABI::buildDynamicCastToVoidInfo(CIRGenFunction &CGF,
+                                                QualType SrcRecordTy) {
+  mlir::Type vtableElemTy;
```
Adding an attribute seems not ideal, it just papers over the fact that we need more information available in LoweringPrepare. See specific comments below.

Lancern replied:
Well, the problem here is dyn-cast-to-void requires completely different set of ABI-specific information than a normal downcast or sidecast. So if you take the attribute approach (i.e. you throw all interesting ABI-specific information into an attribute and attach that attribute to the operation) you can't avoid having two different kinds of attributes.

AST query during lowering prepare seems more promising in this situation. Both `#cir.dyn_cast_info` and `#cir.dyn_cast_to_void_info` are computed from the AST, thus we can delay their computation until lowering prepare. This further allows us to remove `#cir.dyn_cast_info` and `#cir.dyn_cast_to_void_info` completely and make `cir.dyn_cast` clearer in its assembly form.

bcardosolopes replied:
I think there might be value in keeping `#cir.dyn_cast_info` for now, it's already there and it encodes useful information making it easier to lower. More specifically, for keeping a direct reference to the symbols, it might be really handy for optimizations: e.g. it allows us to identify symbol usage/dependency, which is useful for deleting unused code (or globals).

`#cir.dyn_cast_to_void_info` is a different story IMO and could be implemented by querying the AST. I previously gave the suggestion of adding another kind, such as `DCK_PtrCastToVoid`, but couldn't we just look at the op result type to check whether it's a void pointer? I see this in `CIROps.td` definition:

```
let results = (outs StructPtr:$result);
```

Should this be a `StructOrVoidPtr` instead?

Lancern replied:
> but couldn't we just look at the op result type to check whether it's a void pointer?

Yes of course this is possible. Instead of adding more stuff to the existing one, what about a new operation specifically for dyn-cast-to-void? Something like `cir.dyn_cast_to_void`.

bcardosolopes replied:
I'd prefer we encapsulate into just one operation, looks like this could be solved by adding a simpler version of your attribute: an `UnitAttr:$relative_layout`. This is all that lowering should look at, it should provide ABI answers just as accurate as during CIRGen.

Lancern replied:
Updated.

bcardosolopes commented on clang/lib/CIR/CodeGen/CIRGenItaniumCXXABI.cpp:
```diff
@@ -2288,6 +2288,23 @@ static mlir::cir::FuncOp getItaniumDynamicCastFn(CIRGenFunction &CGF) {
+                                                QualType SrcRecordTy) {
+  mlir::Type vtableElemTy;
+  uint64_t vtableElemAlign;
+  if (CGM.getItaniumVTableContext().isRelativeLayout()) {
```
This could be encoded in the module level during CIRGen and queried in LoweringPrepare instead.

Lancern replied:
Updated.

bcardosolopes commented on clang/lib/CIR/CodeGen/CIRGenItaniumCXXABI.cpp:
```diff
@@ -2288,6 +2288,23 @@ static mlir::cir::FuncOp getItaniumDynamicCastFn(CIRGenFunction &CGF) {
+    vtableElemTy = CGF.getBuilder().getSInt32Ty();
+    vtableElemAlign = 4;
+  } else {
+    vtableElemTy = CGF.ConvertType(CGF.getContext().getPointerDiffType());
```
This is an AST query, which is also availble in LoweringPrepare.

Lancern replied:
Updated.

bcardosolopes commented on clang/lib/CIR/CodeGen/CIRGenItaniumCXXABI.cpp:
```diff
@@ -2288,6 +2288,23 @@ static mlir::cir::FuncOp getItaniumDynamicCastFn(CIRGenFunction &CGF) {
+    vtableElemAlign = CGF.getPointerAlign().getQuantity();
+  }
+
+  return mlir::cir::DynamicCastToVoidInfoAttr::get(vtableElemTy,
```
If you'd still need a way to differentiate casts to void, you could perhaps add a new `DCK_PtrCastToVoid` kind, or something.

Lancern replied:
Updated.

bcardosolopes reviewed (changes requested):
This looks mostly great, thanks for working on this!

Lancern commented:
Rebased onto the latest `main`.

Lancern commented:
Rebased onto the latest `main`.

BTW, I found a bug in the previous related patch. The `cir.dyn_cast` operation is lowered to a `cir.ternary` that guards against the source pointer being null. I got the true branch and the false branch wrong. Fixed it inline together with the latest rebase and improvements.

bcardosolopes commented on clang/include/clang/CIR/Dialect/IR/CIROps.td:
```diff
@@ -163,25 +164,43 @@ def DynamicCastOp : CIR_Op<"dyn_cast"> {
+      (`,` qualified($info)^)?
+      (`,` `relative_layout` $relative_layout^)?
+    `)`
+    `->` qualified(type($result)) attr-dict
```
We should start adding new operations without parenthesis, but no need to fix it for this PR, just a thing we should start doing in general.

bcardosolopes commented on clang/include/clang/CIR/Dialect/IR/CIROps.td:
```diff
@@ -163,25 +164,43 @@ def DynamicCastOp : CIR_Op<"dyn_cast"> {
+    `(`
+      $kind `,` $src `:` type($src)
+      (`,` qualified($info)^)?
+      (`,` `relative_layout` $relative_layout^)?
```
Please add some CIRGen testing for this keyword!

Lancern replied:
Updated.

I have to update the syntax a bit: I removed the comma before the `relative_layout` keyword, or the ODS parser would refuse to parse if `$info` is not given but `relative_layout` is present.

bcardosolopes reviewed (approved):
Thanks for fixing the correctness issue too. It's just missing a test now
